File: 0872-Leaf-Similar Trees/0872-Leaf-Similar Trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int> leaves1, leaves2;
        dfs(root1, leaves1);
        dfs(root2, leaves2);
        return leaves1 == leaves2;
    }
    
private:
    void dfs(TreeNode* root, vector<int>& leaves) {
        if (root == nullptr) {
            return;
        }
        
        if (root->left == nullptr && root->right == nullptr) {
            leaves.push_back(root->val);
        }
        else {
            dfs(root->left, leaves);
            dfs(root->right, leaves);
        }        
    }
};

File: 0801-0900/0872-Leaf-Similar Trees/0872-Leaf-Similar Trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def leafSimilar(self, root1: TreeNode, root2: TreeNode) -> bool:
        def dfs(root):
            if root:
                if root.left is None and root.right is None:
                    yield root.val
                else:
                    yield from dfs(root.left)
                    yield from dfs(root.right)

        return list(dfs(root1)) == list(dfs(root2))
